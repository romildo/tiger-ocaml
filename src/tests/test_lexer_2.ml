module L = Lexing

let scan_string s =
  let lexbuf = L.from_string s in
  let rec go () =
    let tok = Lexer.token lexbuf in
    Format.printf
      "%a %s\n%!"
      Location.print_loc (Location.curr_loc lexbuf)
      (Lexer.show_token tok);
    match tok with
    | Parser.EOF -> ()
    | _ -> go ()
  in
  go ()

let lex_string s =
  let lexbuf = L.from_string s in
  let rec go tokens =
    let tok = Lexer.token lexbuf in
    let tokens' = tok :: tokens in
    match tok with
    | Parser.EOF -> List.rev tokens'
    | _ -> go tokens'
  in
  go []

let (>::) title input =
  Printf.printf "=======================================\n";
  Printf.printf "test: %s\n" title;
  Printf.printf "---------------------------------------\n";
  print_endline input;
  Printf.printf "---------------------------------------\n";
  scan_string input;
  Printf.printf "%!"

let test () =
  "eof"              >:: "";
  "spaces"           >:: "  \t\n\n\n  ";
  "integer literals" >:: "0 1342 -56 +32 98fim 139015.007";
  "string literals"  >:: {|"text" "abc\tEND!"|};
  "string literals"  >:: "\"tab \\t; nl \\n; quote \\\"; slash \\\\; control i \^I; 065 \065; empty \\  \t \\;\"";
  "identifier"       >:: "altura";
  "comment"          >:: "/* ignore me */";
  "nested comments"  >:: "/* level 1\n /* level 2\n  /* level 3 */ level 2 /* level 3 */\n level 2 */ level 1 */"(*);
  "unterminated comment" >:: "/* level 1 /* level 2 /* level 3 */ level 2 /* level 3 */" *)

let () = test ()
