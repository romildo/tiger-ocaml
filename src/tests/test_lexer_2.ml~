module Trm = ANSITerminal

module L = Lexing

let scan_string s =
  let lexbuf = L.from_string s in
  let rec go () =
    let tok = Lexer.token lexbuf in
    Format.printf
      "%a %s\n%!"
      Location.print_loc (Location.curr_loc lexbuf)
      (Lexer.show_token tok);
    match tok with
    | Parser.EOF -> ()
    | _ -> go ()
  in
  go ()

let lex_string s =
  let lexbuf = L.from_string s in
  let rec go tokens =
    let tok = Lexer.token lexbuf in
    let tokens' = tok :: tokens in
    match tok with
    | Parser.EOF -> List.rev tokens'
    | _ -> go tokens'
  in
  go []

let (>::) title input =
  Trm.print_string [Trm.blue] "test ";
  Trm.print_string [Trm.green] title;
  Trm.print_string [Trm.blue] "\ninput:\n";
  Trm.print_string [Trm.default] input;
  Trm.print_string [Trm.blue] "\noutput:\n";
  scan_string input;
  Printf.printf "\n%!"

let test () =
  "eof" >:: "";
  "spaces" >:: "  \t\n\n\n  ";
  "integer literals" >:: "1342";
  "string literals" >:: "\"text\"";
  "identifier" >:: "altura";
  "block comments" >:: "/* ignore me */"

let () = test ()
