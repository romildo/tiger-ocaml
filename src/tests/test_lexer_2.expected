=======================================
test: eof
---------------------------------------

---------------------------------------
:1.0-1.0 Parser.EOF
=======================================
test: spaces
---------------------------------------
  	


  
---------------------------------------
:4.2-4.2 Parser.EOF
=======================================
test: integer literals
---------------------------------------
0 1342 -56 +32 98fim 139015.007
---------------------------------------
:1.0-1.1 (Parser.INT 0)
:1.2-1.6 (Parser.INT 1342)
:1.7-1.8 Parser.MINUS
:1.8-1.10 (Parser.INT 56)
:1.11-1.12 Parser.PLUS
:1.12-1.14 (Parser.INT 32)
:1.15-1.17 (Parser.INT 98)
:1.17-1.20 (Parser.ID "fim")
:1.21-1.27 (Parser.INT 139015)
:1.27-1.28 Parser.DOT
:1.28-1.31 (Parser.INT 7)
:1.31-1.31 Parser.EOF
=======================================
test: string literals
---------------------------------------
"text" "abc\tEND!"
---------------------------------------
:1.0-1.6 (Parser.STR "text")
:1.7-1.18 (Parser.STR "abc\tEND!")
:1.18-1.18 Parser.EOF
=======================================
test: string literals
---------------------------------------
"tab \t; nl \n; quote \"; slash \\; control i \^I; 065 A; empty \  	 \;"
---------------------------------------
:1.0-1.72 (Parser.STR "tab \t; nl \n; quote \"; slash \\; control i \t; 065 A; empty ;")
:1.72-1.72 Parser.EOF
=======================================
test: identifier
---------------------------------------
altura
---------------------------------------
:1.0-1.6 (Parser.ID "altura")
:1.6-1.6 Parser.EOF
=======================================
test: comment
---------------------------------------
/* ignore me */
---------------------------------------
:1.15-1.15 Parser.EOF
=======================================
test: nested comments
---------------------------------------
/* level 1
 /* level 2
  /* level 3 */ level 2 /* level 3 */
 level 2 */ level 1 */
---------------------------------------
:4.22-4.22 Parser.EOF
